`showMvnDepsTree` hangs

The project has a large number of dependencies, but for some reasons, `showMvnDepsTree` started to hang without ever returning. I'm pretty sure it used to work before.


[stacktrace.txt](https://github.com/user-attachments/files/25266330/stacktrace.txt)

```
> head -n50 stacktrace.txt
2026-02-12 16:04:08
Full thread dump OpenJDK 64-Bit Server VM (21.0.9+10-LTS mixed mode, sharing):

Threads class SMR info:
_java_thread_list=0x00007f8e00001c20, length=41, elements={
0x00007f8ea40302c0, 0x00007f8ea414ac90, 0x00007f8ea414c200, 0x00007f8ea414d980,
0x00007f8ea414efc0, 0x00007f8ea4150560, 0x00007f8ea4152120, 0x00007f8ea41537e0,
0x00007f8ea41602d0, 0x00007f8ea4163520, 0x00007f8ea4876610, 0x00007f8ea49bdc90,
0x00007f8ea49be810, 0x00007f8ea4a43c30, 0x00007f8ea4a4d710, 0x00007f8ea4b02d90,
0x00007f8ea4b05880, 0x00007f8ea526d9d0, 0x00007f8ea526e9c0, 0x00007f8ea5219ff0,
0x00007f8ea52415b0, 0x00007f8ea52426c0, 0x00007f8ea5273a10, 0x00007f8ea5274870,
0x00007f8ea52758f0, 0x00007f8d581af1d0, 0x00007f8d5415ab40, 0x00007f8d3c02a2b0,
0x00007f8d4000a1b0, 0x00007f8d34001c70, 0x00007f8d38001e50, 0x00007f8ea57ba990,
0x00007f8d6c021320, 0x00007f8d5c050460, 0x00007f8ea4fcf230, 0x00007f8d44061810,
0x00007f8d14005f10, 0x00007f8d20001910, 0x00007f8d1c00f750, 0x00007f8d501ca470,
0x00007f8e00000fc0
}

"main" #1 [25984] prio=5 os_prio=0 cpu=14791,10ms elapsed=25,45s tid=0x00007f8ea40302c0 nid=25984 runnable  [0x00007f8eaa08a000]
   java.lang.Thread.State: RUNNABLE
        at scala.collection.IterableOnceOps.exists(IterableOnce.scala:661)
        at scala.collection.IterableOnceOps.exists$(IterableOnce.scala:337)
        at scala.collection.AbstractIterable.exists(Iterable.scala:946)
        at coursier.core.DependencySet.containsModule(DependencySet.scala:134)
        at coursier.graph.DependencyTree$Node$$anonfun$1.applyOrElse(DependencyTree.scala:162)
        at coursier.graph.DependencyTree$Node$$anonfun$1.applyOrElse(DependencyTree.scala:161)
        at scala.collection.immutable.List.collect(List.scala:269)
        at scala.collection.immutable.List.collect(List.scala:254)
        at coursier.graph.DependencyTree$Node.children(DependencyTree.scala:161)
        at coursier.util.Print$.$anonfun$dependencyTree0$11(Print.scala:235)
        at coursier.util.Print$$$Lambda/0x00007f8e24aabdb0.apply(Unknown Source)
        at coursier.util.Tree.$anonfun$customRender$3(Tree.scala:49)
        at coursier.util.Tree.$anonfun$customRender$3$adapted(Tree.scala:43)
        at coursier.util.Tree$$Lambda/0x00007f8e24aad320.apply(Unknown Source)
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:633)
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:337)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1328)
        at coursier.util.Tree.recursivePrint$1(Tree.scala:43)
        at coursier.util.Tree.$anonfun$customRender$3(Tree.scala:49)
        at coursier.util.Tree.$anonfun$customRender$3$adapted(Tree.scala:43)
        at coursier.util.Tree$$Lambda/0x00007f8e24aad320.apply(Unknown Source)
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:633)
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:337)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1328)
        at coursier.util.Tree.recursivePrint$1(Tree.scala:43)
        at coursier.util.Tree.$anonfun$customRender$3(Tree.scala:49)
        at coursier.util.Tree.$anonfun$customRender$3$adapted(Tree.scala:43)
        at coursier.util.Tree$$Lambda/0x00007f8e24aad320.apply(Unknown Source)
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:633)
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:337)
```

The `main` branch already contains coursier 2.1.25-M25 (https://github.com/com-lihaoyi/mill/pull/7098). Unfortunatelly, the issue is still there.

Also the `OutOfMemorError` with `--inverse` (https://github.com/com-lihaoyi/mill/issues/6823#issuecomment-3896856514) is still present:

```console
> mill cloud.vaadin.showMvnDepsTree --inverse
Mill version 1.1.6-100-f398ed is different than configured for this directory!
Configured version is 1.1.6 (../mill-workspace/build.mill)
...
579] cloud.vaadin.showMvnDepsTree
Exception in thread "FileToStreamTailerThread" Exception in thread "prompt-logger-updater-thread" Exception in thread "FileToStreamTailerThread" java.lang.OutOfMemoryError: Java heap space
        at java.base/java.nio.CharBuffer.wrap(CharBuffer.java:443)
        at java.base/sun.nio.cs.StreamDecoder.implRead(StreamDecoder.java:306)
        at java.base/sun.nio.cs.StreamDecoder.read(StreamDecoder.java:186)
        at java.base/java.io.InputStreamReader.read(InputStreamReader.java:183)
        at java.base/java.io.BufferedReader.fill(BufferedReader.java:166)
        at java.base/java.io.BufferedReader.readLine(BufferedReader.java:333)
        at java.base/java.io.BufferedReader.readLine(BufferedReader.java:400)
        at mill.client.FileToStreamTailer.run(FileToStreamTailer.scala:40)
java.lang.OutOfMemoryError: Java heap space
java.lang.OutOfMemoryError: Java heap space
        at mill.launcher.DaemonRpc$ServerToClient$GetTerminalDims$.apply(DaemonRpc.scala:49)
        at mill.daemon.MillMain0$.$anonfun$20(MillMain0.scala:757)
        at mill.daemon.MillMain0$$$Lambda/0x000000008f22a0e0.apply(Unknown Source)
        at mill.internal.PromptLogger.$init$$$anonfun$2$$anonfun$1(PromptLogger.scala:84)
        at mill.internal.PromptLogger.$init$$$anonfun$2$$anonfun$adapted$1(PromptLogger.scala:97)
        at mill.internal.PromptLogger$$Lambda/0x000000008f24ef98.apply(Unknown Source)
        at mill.api.daemon.StartThread$.$anonfun$1(SpawnThread.scala:5)
        at mill.api.daemon.StartThread$$$Lambda/0x000000008f15c4e8.run(Unknown Source)
        at java.base/java.lang.Thread.runWith(Thread.java:1487)
        at java.base/java.lang.Thread.run(Thread.java:1474)
578/579] mill cloud.vaadin.showMvnDepsTree --inverse 156s
java.lang.Exception: fatal exception occurred: java.lang.OutOfMemoryError: Java heap space
        at scala.collection.IterableOps.withFilter(Iterable.scala:422)
        at scala.collection.IterableOps.withFilter$(Iterable.scala:137)
        at scala.collection.AbstractIterable.withFilter(Iterable.scala:946)
        at coursier.graph.ReverseModuleTree$Node.actualDependees(ReverseModuleTree.scala:280)
        at coursier.graph.ReverseModuleTree$Node.dependees(ReverseModuleTree.scala:296)
        at coursier.util.Print$.$anonfun$dependencyTree0$8(Print.scala:211)
        at coursier.util.Print$.$anonfun$dependencyTree0$8$adapted(Print.scala:210)
        at coursier.util.Print$$$Lambda/0x0000000090926ab8.apply(Unknown Source)
        at scala.collection.immutable.List.noneIn$1(List.scala:509)
        at scala.collection.immutable.List.filterCommon(List.scala:575)
        at scala.collection.immutable.List.filter(List.scala:496)
        at scala.collection.immutable.List.filter(List.scala:496)
        at coursier.util.Print$.$anonfun$dependencyTree0$7(Print.scala:210)
        at coursier.util.Print$$$Lambda/0x0000000090923888.apply(Unknown Source)
        at coursier.util.Tree.$anonfun$customRender$3(Tree.scala:49)
        at coursier.util.Tree.$anonfun$customRender$3$adapted(Tree.scala:43)
        at coursier.util.Tree$$Lambda/0x0000000090924df8.apply(Unknown Source)
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:633)
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:337)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1328)
        at coursier.util.Tree.recursivePrint$1(Tree.scala:43)
        at coursier.util.Tree.$anonfun$customRender$3(Tree.scala:49)
        at coursier.util.Tree.$anonfun$customRender$3$adapted(Tree.scala:43)
        at coursier.util.Tree$$Lambda/0x0000000090924df8.apply(Unknown Source)
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:633)
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:337)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1328)
        at coursier.util.Tree.recursivePrint$1(Tree.scala:43)
        at coursier.util.Tree.$anonfun$customRender$3(Tree.scala:49)
        at coursier.util.Tree.$anonfun$customRender$3$adapted(Tree.scala:43)
        at coursier.util.Tree$$Lambda/0x0000000090924df8.apply(Unknown Source)
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:633)
Caused by: java.lang.OutOfMemoryError: Java heap space
        ... 32 more

mill.api.daemon.Result$SerializedException: fatal exception occurred: java.lang.OutOfMemoryError: Java heap space
        at scala.collection.IterableOps.withFilter(Iterable.scala:422)
        at scala.collection.IterableOps.withFilter$(Iterable.scala:137)
        at scala.collection.AbstractIterable.withFilter(Iterable.scala:946)
        at coursier.graph.ReverseModuleTree$Node.actualDependees(ReverseModuleTree.scala:280)
        at coursier.graph.ReverseModuleTree$Node.dependees(ReverseModuleTree.scala:296)
        at coursier.util.Print$.$anonfun$dependencyTree0$8(Print.scala:211)
        at coursier.util.Print$.$anonfun$dependencyTree0$8$adapted(Print.scala:210)
        at coursier.util.Print$$$Lambda/0x0000000090926ab8.apply(Unknown Source)
        at scala.collection.immutable.List.noneIn$1(List.scala:509)
        at scala.collection.immutable.List.filterCommon(List.scala:575)
        at scala.collection.immutable.List.filter(List.scala:496)
        at scala.collection.immutable.List.filter(List.scala:496)
        at coursier.util.Print$.$anonfun$dependencyTree0$7(Print.scala:210)
        at coursier.util.Print$$$Lambda/0x0000000090923888.apply(Unknown Source)
        at coursier.util.Tree.$anonfun$customRender$3(Tree.scala:49)
        at coursier.util.Tree.$anonfun$customRender$3$adapted(Tree.scala:43)
        at coursier.util.Tree$$Lambda/0x0000000090924df8.apply(Unknown Source)
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:633)
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:337)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1328)
        at coursier.util.Tree.recursivePrint$1(Tree.scala:43)
        at coursier.util.Tree.$anonfun$customRender$3(Tree.scala:49)
        at coursier.util.Tree.$anonfun$customRender$3$adapted(Tree.scala:43)
        at coursier.util.Tree$$Lambda/0x0000000090924df8.apply(Unknown Source)
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:633)
        at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:337)
        at scala.collection.AbstractIterator.foreach(Iterator.scala:1328)
        at coursier.util.Tree.recursivePrint$1(Tree.scala:43)
        at coursier.util.Tree.$anonfun$customRender$3(Tree.scala:49)
        at coursier.util.Tree.$anonfun$customRender$3$adapted(Tree.scala:43)
        at coursier.util.Tree$$Lambda/0x0000000090924df8.apply(Unknown Source)
        at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:633)
Caused by: mill.api.daemon.Result$SerializedException: Java heap space
        ... 32 more

Mill launcher failed. See out/mill-launcher/log for details.
```

--- a/libs/javalib/src/mill/javalib/JavaModule.scala
+++ b/libs/javalib/src/mill/javalib/JavaModule.scala
@@ -40,7 +40,6 @@ import os.Path
 
 import java.io.File
 import scala.util.chaining.scalaUtilChainingOps
-import scala.util.matching.Regex
 
 /**
  * Core configuration required to compile a single Java module

--- a/libs/javalib/src/mill/javalib/JavaModule.scala
+++ b/libs/javalib/src/mill/javalib/JavaModule.scala
@@ -1463,18 +1463,21 @@ trait JavaModule
             .filter(dep => matchers.exists(matcher => matcher.matches(dep.module))).toSeq
       }
 
-      val tree = coursier.util.Print.dependencyTree(
+      val tree = coursier.util.Print.dependencyTree0(
         resolution = resolution,
         roots = roots,
         printExclusions = false,
-        reverse = if (whatDependsOn.isEmpty) inverse else true
+        reverse = if (whatDependsOn.isEmpty) inverse else true,
+        // Fix issue: https://github.com/com-lihaoyi/mill/issues/6823
+        // see also comment: https://github.com/coursier/coursier/pull/3671#issuecomment-4752734517
+        reverseDeduplicateNodes = true
       )
 
       // Filter the output, so that the special organization and version used for Mill's own modules
       // don't appear in the output. This only leaves the modules' name built from millModuleSegments.
       val processedTree = tree
         .replace(s"${JavaModule.internalOrg.value}:", "")
-        .pipe(JavaModule.removeInternalVersionRegex.replaceAllIn(_, "$1"))
+        .replace(s":${JavaModule.internalVersion}", "")
 
       processedTree
     }
@@ -1852,9 +1855,11 @@ object JavaModule {
 
   private[mill] def internalOrg = coursier.core.Organization("mill-internal")
   private[mill] def internalVersion = "0+mill-internal"
-
-  private lazy val removeInternalVersionRegex =
-    (":" + Regex.quote(JavaModule.internalVersion) + "(\\w*$|\\n)").r
+//  private lazy val removeInternalVersionRegex =
+//    (
+//      ":" + Regex.quote(JavaModule.internalVersion) +
+//        "(\\w*(" + Regex.quote(dedupTreeSuffix) + ")?$|\\n)"
+//    ).r
 
 }
 

--- a/libs/javalib/src/mill/javalib/JavaModule.scala
+++ b/libs/javalib/src/mill/javalib/JavaModule.scala
@@ -1854,11 +1854,6 @@ object JavaModule {
 
   private[mill] def internalOrg = coursier.core.Organization("mill-internal")
   private[mill] def internalVersion = "0+mill-internal"
-//  private lazy val removeInternalVersionRegex =
-//    (
-//      ":" + Regex.quote(JavaModule.internalVersion) +
-//        "(\\w*(" + Regex.quote(dedupTreeSuffix) + ")?$|\\n)"
-//    ).r
 
 }
 

